### PR TITLE
Add left panel hierarchy support

### PR DESCRIPTION
**Pull request by igor-dv**
Issue: 
[Substories/Hierarchy #151](https://github.com/storybooks/storybook/issues/151)

## What I did
I've added the option to visualize the "namespace syntax" of stories as a hierarchy in the stories panel (left panel).

For example, by providing stories like these:
```javascript
storiesOf('web.component.Button', module)
  .add('One', ...)
  .add('Two', ...)
  .add('Three', ...)

storiesOf('web.component.Link', module)
  .add('One', ...)
  .add('Two', ...)
  .add('Three', ...)
```

The left panel will be like this:
```
web
|  component
|  |  Button
|  |      One    <-- if this is selected, other components are collapsed
|  |      Two
|  |      Three
|  |  Link
```

You'll be able to control it by defining the new "leftPanelHierarchy" prop of the uiOptions as a true (by default it's false).

## How to test
You can run the storybook/lib/ui/example

### CODE REVIEW

**Comment by UsulPro** [comment]
@igor-dv Thanks for this implementation! 
UI and API looks great to me! :clap:

I played with this solution and found some issues:

1. stories with hierarchy are located at the end of stories list even if I add them in the middle

2. It creates two separate items in the stories list if I `.add` stories in the middle of hierarchy path. e.g.:

```js
storiesOf('Atoms.Moleculs', module)
  .add('with text', () => <Button>Hello Button</Button>)
  .add('with some emoji', () => <Button>😀 😎 👍 💯</Button>)


storiesOf('Atoms.Moleculs.Cells', module)
  .add('with text2', () => <Button>Hello Button</Button>)
  .add('with some emoji2', () => <Button>😀 😎 👍 💯</Button>)
```

I've got such menu:
image

expected behavior: show only one item with stories and a substory item

3. It doesn't work with Knobs. I tried to change some values - no response.

4. I played with some nested levels and faced that in some cases it opens not the first item when I select the storyKind.

<summary>
the way to reproduce
<details>

```js
storiesOf('Atoms.Moleculs.Cells.simple', module)
  .add('with text', () => <Button>Hello Button</Button>)
  .add('with some emoji', () => <Button>😀 😎 👍 💯</Button>)

storiesOf('Atoms.Moleculs.Cells.more', module)
  .add('with text2', () => <Button>Hello Button</Button>)
  .add('with some emoji2', () => <Button>😀 😎 👍 💯</Button>)

storiesOf('Atoms.Moleculs', module)
  .add('with text', () => <Button>Hello Button</Button>)
  .add('with some emoji', () => <Button>😀 😎 👍 💯</Button>)

storiesOf('Atoms.Moleculs.Cells', module)
  .add('with text2', () => <Button>Hello Button</Button>)
  .add('with some emoji2', () => <Button>😀 😎 👍 💯</Button>)

```
When I click `Atoms` first time it opens all nested items this way:

image

possible it's overwhelming case thought :)


</details>

<summary>

### My suggestion

In general, I like this solution. In order to succeed with this feature don't you mind to collaborate with [storybook-addon-chapters](https://github.com/sm-react/storybook-chapters) wich provides hierarchy as well but in another way? And possible with other developers who'd like to jump to this feature. So we can create a separate branch for this feature and continue the work.

**Review by usulpro — CHANGES_REQUESTED** [review]
see some issues above

**Comment by alexandrebodin** [comment]
My bad I clicked the wrong button xD

**Comment by alexandrebodin** [comment]
Full disclosure.

I'm working a moving the ui to redux instead of mantra. While doing this I am making a hierarchical view to. I think we should merge @igor-dv PR because my work is not ready yet. 

I'll open a PR so @igor-dv can look at it and the kirks I had to handle to make the hierarchy work has I wanted (Hoping this will help you with fix bugs :)) 

If you want to have a look I'm opening a PR https://github.com/storybooks/storybook/pull/1334

**Comment by igor-dv** [comment]
@UsulPro , thanks for your review.

Regarding the issues:
1. Fixed
2. I don't know if it's a desired behavior . I assume that will be provided a "fully qualified" component name.
For example in "Atoms.Molecules" , "Atoms" is a namespace and "Molecules" is a component name. In "Atoms.Molecules.Cells" - "Atoms.Molecules" is a namespace and "Cells" is component name.. 
Would like to here more people regarding this, but maybe you are right 
3. I checked it with Knobs and it worked for me (added a few examples in **cra-kitchen-sink**)
4. After some fixes it looks ok =)

Regarding the **storybook-addon-chapters**. I've considered it as an alternative for the projects in a company I work, but it didn't meet our needs.
1. It always changes the context of the stories in the left panel
2. It adds things to the preview section

In general, I think that the left panel should look like most of the IDEs files explorer (or how it's called..)

Though I am big fan of the plugins idea, Storybok addons api doesn't allow us to create a proper addon for the left panel.. 

Anyway, I think it's a great plugin, and would like to help with it. 

After a bit UI changes, it looks like this now:
storybook_hierarchy

**Comment by igor-dv** [comment]
storybook_hierarchy_toggle

**Comment by shilman** [comment]
@igor-dv my guess is that "toggle story hierarchy" is gratuitous. My guess is that we release this as a backwards-compatible feature in 3.2 and then, assuming lots of people use it, we just make it the default in 4.0.

**Comment by shilman** [comment]
@igor-dv I think it would be nice to be able to pass a function that constructs the hierarchy from the story name. So by default that would be `(storyName) => storyName.split(".")` but that it could be overridden in the configuration.

Use cases:
1. Sketch symbols use the naming convention `a/b/c/SymbolName`, and I've considered exporting these to Storybook as part of a design system
2. I've also done things like `kindName-storyName_deviceSize` and would probably override the default in some collaboration-related features I've developed in private addons.

**Comment by igor-dv** [comment]
@shilman 
Regarding the splitting - great idiea. I felt it's a bit hardcoded. So i can provide it as a parameter in the 
`uiOptions.leftPanelHierarchy`

like by default it's a boolean - `uiOptions.leftPanelHierarchy = false`

but in more expanded way it's 
```javascript
uiOptions.leftPanelHierarchy = {
  enabled: true,
  split: story => (storyName) => storyName.split(".")
}
```

WDYT ?

**Comment by shilman** [comment]
@igor-dv another idea (don't know if it's a good one):

Existing behavior (default):
```
uiOptions = {
  storyHierarchy: (storyName) => storyName
}
```

New behavior:
```
uiOptions = {
  storyHierarchy: (storyName) => storyName.split('.')
}
```

**Comment by alexandrebodin** [comment]
I'm picky but I'm not sur about using 'Kind1.Kind2' 
I would suggest a path like syntax 'Kind1/Kind2' to communicate the intent a bit more.

In my opinion `.` might be uses widely for any other meaning. and `/` is already used by users to have a visual clue of the separation. This would bring direct benefits to those already using this standard way of declaring URI don't you think ?

**Comment by shilman** [comment]
@alexandrebodin I don't have a strong opinion about this, but `/` would also be my preferred choice (see the Sketch comment above)

**Comment by igor-dv** [comment]
@shilman 
```
uiOptions = {
  storyHierarchy: (storyName) => [storyName]
}
``` 
That works.. but toggling became harder to handle.. I need to add something to state.. 

@alexandrebodin 
If we will provide the splitting function as an option, it will be easier to tweak this..
WDYT ?

**Comment by shilman** [comment]
@igor-dv why do we need toggling?

Also, this may be out of scope, but how hard would it be to replace the left panel with a generic tree widget, such as (react-treebeard](https://alexcurtis.github.io/react-treebeard/) with custom styles and key-bindings.

I think Storybook's custom list UI kind of made sense when it was a flat list, but now that we have hierarchy, there are already common UI's for dealing with this that everybody (including non-engineers) are used to using.

File: examples/cra-kitchen-sink/src/stories/index.js [hunk 1]
```diff
@@ -153,3 +153,43 @@ storiesOf('WithEvents', module)
...
+  .add('first', () => <table><tr><td>first table</td></tr></table>)
+  .add('second', () => <table><tr><td>first table</td></tr></table>);
+
+storiesOf('component.Button')
+  .add('first', () => <button>first button</button>)
+  .add('second', () => <button>first second</button>);
+
+// Atomic
+
+storiesOf('Atoms.Molecules.Cells.simple', module)
```
Comment (shilman) [inline examples/cra-kitchen-sink/src/stories/index.js]: I think the order is backwards: Cells.Molecules.Atoms?

File: lib/ui/example/client/provider.js [hunk 2]
```diff
@@ -78,25 +78,62 @@ export default class ReactProvider extends Provider {
     this.api = api;
     this.api.setOptions({
       name: 'REACT-STORYBOOK',
+      sortStoriesByKind: true,
```
Comment (shilman) [inline lib/ui/example/client/provider.js]: "breaking"?

Comment (igor-dv) [inline lib/ui/example/client/provider.js]: As far as I checked - no. But maybe it's already has another meaning.. Anyway the hierarchy is built after the filtering:
image

File: examples/cra-kitchen-sink/src/stories/index.js (hunk 1 above)
Comment (UsulPro) [inline examples/cra-kitchen-sink/src/stories/index.js]: possible we could get our existing kitchen-sink stuff and reorganize in a hierarchical way?

**Comment by igor-dv** [comment]
@shilman 
I just thought toggling will be a nice thing, but you are right, I don't see benefits of having the old style view.

treebeard - interesting, I'll check it..

**Comment by igor-dv** [comment]
I think a better name will be 
```javascript
uiOptions = {
  resolveStoryHierarchyNamespace: (storyName) => [storyName]
}
```

**Comment by shilman** [comment]
@igor-dv how about just `resolveStoryHierarchy` ... or something a little shorter! 😄

**Comment by ndelangen** [comment]
@igor-dv I'm loving this! This is amazing!

**Comment by shilman** [comment]
**NOTE:** merged this into `151-story-hierarchy` branch so @igor-dv can continue working in the storybook repo. 👍

**Pull request merged**